### PR TITLE
Feature/add chatroom

### LINKED ISSUE #61: Share media files
**__Feature Description__**
Allow users to share media files between them.

**__Major User Stories/tasks__**
- [ ] 1
- [ ] 2
- [ ] 3

**__Additional Considerations__**
Include anything else that may be helpful. For example, links to external resources.

### LINKED ISSUE #129: Implement the Chat Room page
Create chatroom components and style them, and add form handlers.

### LINKED ISSUE #130: Create messages store
Create messages slice and actions

### LINKED ISSUE #131: Create the messages service
Create the messages service that includes three functions: getAll, createTextMessage, and createPictureMessage.

### PR DESCRIPTION
**Pull request by KhadidjaArezki**
First of all, I apologize for creating such a large PR. I should have created one when the sending text messages feature was done, but I didn't know that the sending pictures feature will require all this time and effort. I should have known better.
The code is not perfect. I tried to make it as clear as possible, and I added comments. The UseCameraModal is still messy as it's quite large. 
- I refactored the Image modal to be used anywhere.
- I modified the messages router and controller to handle image files.

**TODO**
- There are some problems in displaying picture messages on mobile that need to be fixed, namely, the picture container is smaller than the picture, and pictures taken with camera are distorted because their height shrinks for some reason.
- Redux SerializableStateInvariantMiddleware middleware slows the display of added messages, but those are active during development mode only and not in production builds, and they can be disabled in dev mode too.

closes #129, closes #130, closes #131, and closes #61

### CODE REVIEW

File: client/src/components/message_body/MessageBody.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  const getBinaryStringFromByteArray = (byteArray) => (
+    new Uint8Array(byteArray)
+      .reduce((data, byte) => data + String.fromCharCode(byte), '')
+  )
+
+  /* Convert image byte array into an image url */
+  const imgToDataUrl = () => {
+    const binaryString = getBinaryStringFromByteArray(content.data.data)
+    /* create a Base64-encoded ASCII string from a binary string */
+    const image = btoa(binaryString)
```
Comment (moody2times) [inline client/src/components/message_body/MessageBody.js]: `btoa` has been deprecated in `nodejs` only but vscode is showing a warning for browser side usage. You can find a workaround here in this [article](https://stackoverflow.com/questions/68849233/convert-a-string-to-base64-in-javascript-btoa-and-atob-are-deprecated). This is a requested minor change.

File: client/src/components/messages/Messages.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  /* user from the friends array in the user store      */
+  const currentFriendInfo = friends.find((f) => f.id === currentFriendId)
+
+  /* Pass a ref hook to each message in order to auto scroll */
+  /* to the bottom of each message that is displayed         */
+  const scrollRef = useRef()
+
+  /* Auto scroll to bottom on each successfully-added message */
+  useEffect(() => {
+    scrollRef.current?.scrollIntoView(false, { behavior: 'smooth' })
```
Comment (moody2times) [inline client/src/components/messages/Messages.js:27]: You want to scroll the new message into view but you set it to false here. Am I missing something? This is a question and not necessarily a request for change.

File: client/src/components/new_message_form/NewMessageForm.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import { addTextMessage } from '../../store/features/messages/messagesSlice'
+
+function NewMessageForm() {
+  const dispatch = useDispatch()
+
+  const handleSubmit = (event) => {
+    event.preventDefault()
+    const message = event.target.message.value
+    dispatch(addTextMessage(message))
+    event.target.message.value = '' // eslint-disable-line no-param-reassign
```
Comment (moody2times) [inline client/src/components/new_message_form/NewMessageForm.js:13]: I think this is the second time I am seeing this manual way of resetting form control value. In the future, we can change this. This is not a request for change at this time.

File: client/src/services/messages.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
+import axios from 'axios'
+
+const baseUrl = process.env.REACT_APP_BASE_MESSAGES_URL
+
+const getAll = async (id, token) => {
+  const config = {
```
Comment (moody2times) [inline client/src/services/messages.js]: You can import and reuse the utility function in the config.js for this. This is a request for change.

File: client/src/services/messages.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+const createTextMessage = async (textMessageObject, token) => {
+  const config = {
+    headers: { Authorization: `bearer ${token}` },
+  }
+  const response = await axios.post(baseUrl, textMessageObject, config)
+  return response.data
+}
+
+const createPictureMessage = async (pictureMessageformData, token) => {
+  const config = {
```
Comment (moody2times) [inline client/src/services/messages.js]: You can refactor the config utility function to receive this additional parameter and be DRY. This is a request for change.

**Review by moody2times — CHANGES_REQUESTED** [review]
Well done for your hard work.
I think you can still do more to improve your code commenting. I have left some notes and I am not requesting a change in all cases. Some notes are just notes for future references only. When I request a change, I will make it clear. Some things should be left for after we finish.
Also, please try to rebase your topic branch before pushing to get the latest updates.

File: client/src/components/messages/Messages.js (hunk 2 above)
Comment (KhadidjaArezki) [reply client/src/components/messages/Messages.js:27]: The first param -  `alignToTop `- is a boolean. If it's false, the bottom of the element will be aligned to the bottom of the visible area of the scrollable ancestor. The second param - `scrollIntoViewOptions `- is an object that specifies the scrolling behavior.

File: client/src/components/new_message_form/NewMessageForm.js (hunk 3 above)
Comment (KhadidjaArezki) [reply client/src/components/new_message_form/NewMessageForm.js:13]: Absolutely. I am sure there is a React way to do this.

File: client/src/services/messages.js (hunk 4 above)
Comment (KhadidjaArezki) [reply client/src/services/messages.js]: Will do.

File: client/src/services/messages.js (hunk 5 above)
Comment (KhadidjaArezki) [reply client/src/services/messages.js]: I will try.

File: client/src/components/message_body/MessageBody.js (hunk 1 above)
Comment (KhadidjaArezki) [reply client/src/components/message_body/MessageBody.js]: `btoa` and `atob` are only deprecated in node. I added a fix to get rid of the warning. It seems that the compiler thinks we are in a node env for some reason. I will go back and look at better approaches to fix this later.

**Review by moody2times — APPROVED** [review]
This will do. Well done.

**Pull request merged**